runtime/codert_vm/cnathelp.cpp

```diff
@@ -3122,6 +3122,7 @@ old_slow_jitReportStaticFieldWrite(J9VMThread *currentThread)
 	if (J9_EVENT_IS_HOOKED(vm->hookInterface, J9HOOK_VM_PUT_STATIC_FIELD)) {
 		J9Class *fieldClass = dataBlock->fieldClass;
 		if (J9_ARE_ANY_BITS_SET(fieldClass->classFlags, J9ClassHasWatchedFields)) {
+			U_64 value = *(U_64*)valuePointer;
```

Comment:
Does this need to be marked `volatile` to ensure the compiler doesn't move the read?

Reply:
What do you think needs to be volatile? I strongly suspect that the fact that there's a call between the read and the use would likely prevent the motion.

Given the existing checks before the read, I would have no issue marking valuePointer and it's uses as volatile. I strongly dislike marking locals volatile.

Reply:
Another (not necessarily better) option would be the explciit compiler reordering barrier which could be inserted just before the hook dispatch.

Reply:
I agree the reorder barrier would be sufficient but doesn't it need to be before the call to `initializeClass()` which is where the stack grow may occur

Reply:
Yes, that would make more sense - I'll add one there for the sake of it (still pretty sure it's not necessary as we don't do it in many other places).

Reply:
Added a volatile and a barrier.